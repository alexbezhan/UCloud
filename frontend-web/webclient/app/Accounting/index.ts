import Usage from "Accounting/Usage";

export {Usage};