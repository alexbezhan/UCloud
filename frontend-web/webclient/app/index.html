<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=1">
    <title>UCloud</title>
</head>
<body>
    <noscript>
        JavaScript is required to run this page.
    </noscript>
    <div id="app"></div>
</body>
</html>
