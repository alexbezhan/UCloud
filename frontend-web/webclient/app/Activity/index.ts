import { ActivityReduxObject } from "DefaultObjects";
import { ScrollRequest } from "Scroll";

export enum ActivityType {
    DOWNLOAD = "download",
    UPDATED = "updated",
    DELETED = "deleted",
    FAVORITE = "favorite",
    INSPECTED = "inspected",
    MOVED = "moved"
}

export interface ActivityGroup {
    type: ActivityType;
    newestTimestamp: number;
    numberOfHiddenResults: number | null;
    items: Activity[];
}

export interface Activity {
    type: ActivityType;
    timestamp: number;
    fileId: string;
    username: string;
    originalFilePath: string;
}

export interface ActivityFilter {
    collapseAt?: number;
    type?: ActivityType;
    minTimestamp?: Date;
    maxTimestamp?: Date;
}

export interface FavoriteActivity extends Activity {
    favorite: boolean;
}

export interface MovedActivity extends Activity {
    newName: string;
}


export interface ActivityDispatchProps {
    onMount: () => void;
    fetchActivity: (scroll: ScrollRequest<number>, filter?: ActivityFilter) => void;
    resetActivity: () => void;
    setRefresh: (refresh?: () => void) => void;
    updateFilter: (filter: Partial<ActivityFilter>) => void;
}

export interface ActivityOwnProps {
}

export type ActivityProps = ActivityReduxObject & ActivityDispatchProps & ActivityOwnProps;
