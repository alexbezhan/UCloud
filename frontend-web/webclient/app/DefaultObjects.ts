import {SidebarOption, Page} from "Types";
import {Status} from "Navigation";
import {Analysis, DetailedApplicationSearchReduxState, RunsSortBy} from "Applications";
import {File, DetailedFileSearchReduxState, SortOrder} from "Files";
import {DashboardStateProps} from "Dashboard";
import {Notification} from "Notifications";
import {Upload} from "Uploader";
import {Activity, ActivityGroup, ActivityFilter} from "Activity";
import {Reducer} from "redux";
import {SimpleSearchStateProps} from "Search";
import * as ApplicationRedux from "Applications/Redux";
import * as AccountingRedux from "Accounting/Redux";
import {defaultAvatar} from "UserSettings/Avataaar";
import {SidebarPages} from "ui-components/Sidebar";
import {ScrollResult} from "Scroll/Types";
import * as ProjectRedux from "Project/Redux";

export const DefaultStatus: Status = {
    title: "No Issues",
    level: "NO ISSUES",
    body: "The system is running as intended."
};

export enum KeyCode {
    ENTER = 13,
    ESC = 27
}

export const emptyPage: Page<any> = {items: [], itemsPerPage: 25, itemsInTotal: 0, pageNumber: 0, pagesInTotal: 0};

export enum SensitivityLevel {
    "INHERIT" = "Inherit",
    "PRIVATE" = "Private",
    "CONFIDENTIAL" = "Confidential",
    "SENSITIVE" = "Sensitive"
}

export type Sensitivity = keyof typeof SensitivityLevel;

export enum SensitivityLevelMap {
    INHERIT = "INHERIT",
    PRIVATE = "PRIVATE",
    CONFIDENTIAL = "CONFIDENTIAL",
    SENSITIVE = "SENSITIVE"
}

export interface ComponentWithLoadingState {
    loading: boolean
    error?: string
}

export interface ComponentWithPage<T> extends ComponentWithLoadingState {
    page: Page<T>
}

export interface ComponentWithScroll<Item, OffsetType> extends ComponentWithLoadingState {
    scroll?: ScrollResult<Item, OffsetType>
}

export interface ResponsiveReduxObject {
    mediaType: string
    orientation: string
    lessThan: Record<string, boolean>
    greaterThan: Record<string, boolean>
    is: Record<string, boolean>
}

export const initResponsive = (): ResponsiveReduxObject => ({
    mediaType: "",
    orientation: "",
    lessThan: {},
    greaterThan: {},
    is: {},
});

export interface FilePreviewReduxState {
    file?: File
    error?: string
}

export interface FileInfoReduxObject {
    file?: File
    error?: string
    activity: Page<Activity>
    loading: boolean
}

export interface AnalysisReduxObject extends ComponentWithPage<Analysis> {
    sortBy: RunsSortBy
    sortOrder: SortOrder
}

export interface NotificationsReduxObject {
    redirectTo: string
    items: Notification[]
    loading: boolean
    error?: string
}

export interface StatusReduxObject {
    status: Status
    title: string
    page: SidebarPages
    loading: boolean
}

export interface SidebarReduxObject {
    pp: boolean
    options: SidebarOption[]
    kcCount: number
}

export interface HeaderSearchReduxObject {
    prioritizedSearch: HeaderSearchType
    refresh?: () => void
}

export interface RunApplicationReduxObject {

}

export type ActivityReduxObject = ComponentWithScroll<ActivityGroup, number> & ActivityFilter;

export type HeaderSearchType = "files" | "applications" | "projects";

export interface UploaderReduxObject {
    uploads: Upload[]
    visible: boolean
    path: string
    allowMultiple: boolean
    onFilesUploaded: (p: string) => void
    error?: string
    loading: boolean
}

interface LegacyReducers {
    dashboard?: Reducer<DashboardStateProps>
    uploader?: Reducer<UploaderReduxObject>
    status?: Reducer<StatusReduxObject>
    notifications?: Reducer<NotificationsReduxObject>
    analyses?: Reducer<AnalysisReduxObject>
    header?: Reducer<HeaderSearchReduxObject>
    sidebar?: Reducer<SidebarReduxObject>
    activity?: Reducer<ActivityReduxObject>
}

export type Reducers = LegacyReducers & ApplicationRedux.Reducers & AccountingRedux.Reducers;

/* FIXME */
interface LegacyReduxObject {
    dashboard: DashboardStateProps
    uploader: UploaderReduxObject
    status: StatusReduxObject,
    notifications: NotificationsReduxObject
    analyses: AnalysisReduxObject
    header: HeaderSearchReduxObject
    sidebar: SidebarReduxObject
    activity: ActivityReduxObject
    simpleSearch: SimpleSearchStateProps
    detailedFileSearch: DetailedFileSearchReduxState
    detailedApplicationSearch: DetailedApplicationSearchReduxState
    fileInfo: FileInfoReduxObject
    avatar: AvatarReduxObject
    filePreview: FilePreviewReduxState
    responsive?: ResponsiveReduxObject
    project: ProjectRedux.State
    loading?: boolean
}

export type ReduxObject =
    LegacyReduxObject &
    ApplicationRedux.Objects &
    AccountingRedux.Objects;

export const initActivity = (): ActivityReduxObject => ({
    loading: false
});

export const initNotifications = (): NotificationsReduxObject => ({
    items: [],
    loading: false,
    redirectTo: "",
    error: undefined
});

export const initHeader = (): HeaderSearchReduxObject => ({
    prioritizedSearch: "files"
});

export const initStatus = (): StatusReduxObject => ({
    status: DefaultStatus,
    title: "",
    page: SidebarPages.None,
    loading: false
});

export const initDashboard = (): DashboardStateProps => ({
    favoriteFiles: [],
    recentFiles: [],
    recentAnalyses: [],
    notifications: [],
    favoriteLoading: false,
    recentLoading: false,
    analysesLoading: false
});

export function initObject(homeFolder: string): ReduxObject {
    return {
        dashboard: initDashboard(),
        status: initStatus(),
        header: initHeader(),
        notifications: initNotifications(),
        analyses: initAnalyses(),
        sidebar: initSidebar(),
        uploader: initUploads(),
        activity: initActivity(),
        simpleSearch: initSimpleSearch(),
        detailedApplicationSearch: initApplicationsAdvancedSearch(),
        detailedFileSearch: initFilesDetailedSearch(),
        fileInfo: initFileInfo(),
        avatar: initAvatar(),
        filePreview: initFilePreview(),
        project: ProjectRedux.initialState,
        ...ApplicationRedux.init(),
        ...AccountingRedux.init(),
        responsive: undefined,
    }
}


export const initFilePreview = () => ({
    file: undefined,
    error: undefined
});

export type AvatarReduxObject = typeof defaultAvatar & { error?: string };
export const initAvatar = (): AvatarReduxObject => ({...defaultAvatar, error: undefined});

export const initSimpleSearch = (): SimpleSearchStateProps => ({
    files: emptyPage,
    filesLoading: false,
    applications: emptyPage,
    applicationsLoading: false,
    errors: [],
    search: "",
    applicationSearch: initApplicationsAdvancedSearch(),
    fileSearch: initFilesDetailedSearch()
});

export const initAnalyses = (): AnalysisReduxObject => ({
    page: emptyPage,
    loading: false,
    error: undefined,
    sortBy: RunsSortBy.createdAt,
    sortOrder: SortOrder.DESCENDING
});

export const initSidebar = (): SidebarReduxObject => ({
    pp: false,
    kcCount: 0,
    options: []
});

export const initUploads = (): UploaderReduxObject => ({
    path: "",
    uploads: [],
    visible: false,
    allowMultiple: false,
    error: undefined,
    onFilesUploaded: () => null,
    loading: false
});

export const initFileInfo = (): FileInfoReduxObject => ({
    activity: emptyPage,
    loading: false
});

export const initFilesDetailedSearch = (): DetailedFileSearchReduxState => ({
    hidden: true,
    allowFolders: true,
    allowFiles: true,
    fileName: "",
    extensions: new Set(),
    tags: new Set(),
    sensitivities: new Set(),
    createdBefore: undefined,
    createdAfter: undefined,
    modifiedBefore: undefined,
    modifiedAfter: undefined,
    error: undefined,
    loading: false
});

export const initApplicationsAdvancedSearch = (): DetailedApplicationSearchReduxState => ({
    error: undefined,
    loading: false,
    hidden: true,
    appName: "",
    appVersion: "",
    tags: ""
});
