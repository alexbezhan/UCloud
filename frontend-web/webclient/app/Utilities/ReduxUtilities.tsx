import {createStore, combineReducers, Store, AnyAction, Action} from "redux";
import {composeWithDevTools} from 'redux-devtools-extension';
import {ReduxObject, initObject} from "DefaultObjects";
import {responsiveBP} from "ui-components/theme";
import {createResponsiveStateReducer} from "redux-responsive";
import {Cloud} from "Authentication/SDUCloudObject";
import {CONTEXT_SWITCH, USER_LOGOUT, USER_LOGIN} from "Navigation/Redux/HeaderReducer";

export function configureStore(initialObject: Partial<ReduxObject>, reducers, enhancers?): Store<ReduxObject, AnyAction> {
    const combinedReducers = combineReducers<ReduxObject, AnyAction>(reducers);
    const rootReducer = (state: ReduxObject, action: Action): ReduxObject => {
        if ([USER_LOGIN, USER_LOGOUT, CONTEXT_SWITCH].some(it => it === action.type)) {
            state = initObject(Cloud.homeFolder);
        }
        return combinedReducers(state, action);
    };
    return createStore<ReduxObject, AnyAction, {}, {}>(rootReducer, initialObject, composeWithDevTools(enhancers));
}

export const responsive = createResponsiveStateReducer(
    responsiveBP,
    {infinity: "xxl"}
);