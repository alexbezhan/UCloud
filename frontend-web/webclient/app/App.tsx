import * as React from "react";
import * as ReactDOM from "react-dom";
import {Provider} from "react-redux";
import {BrowserRouter} from "react-router-dom";
import fileInfo from "Files/Redux/FileInfoReducer";
import {theme, UIGlobalStyle} from "ui-components";
import {createGlobalStyle, ThemeProvider} from "styled-components";
import {Cloud} from "Authentication/SDUCloudObject";
import {initObject} from "DefaultObjects";
import Core from "Core";
import avatar from "UserSettings/Redux/AvataaarReducer";
import header, {USER_LOGIN, USER_LOGOUT, CONTEXT_SWITCH} from "Navigation/Redux/HeaderReducer";
import status from "Navigation/Redux/StatusReducer";
import applications from "Applications/Redux/BrowseReducer";
import dashboard from "Dashboard/Redux/DashboardReducer";
import sidebar from "Navigation/Redux/SidebarReducer";
import analyses from "Applications/Redux/AnalysesReducer";
import notifications from "Notifications/Redux/NotificationsReducer";
import uploader from "Uploader/Redux/UploaderReducer";
import activity from "Activity/Redux/ActivityReducer";
import simpleSearch from "Search/Redux/SearchReducer";
import detailedFileSearch from "Files/Redux/DetailedFileSearchReducer";
import detailedApplicationSearch from "Applications/Redux/DetailedApplicationSearchReducer";
import filePreview from "Files/Redux/FilePreviewReducer";
import * as AppRedux from "Applications/Redux";
import * as AccountingRedux from "Accounting/Redux";
import {configureStore} from "Utilities/ReduxUtilities";
import {responsiveStoreEnhancer, createResponsiveStateReducer} from 'redux-responsive';
import {responsiveBP, invertedColors} from "ui-components/theme";
import {findAvatar} from "UserSettings/Redux/AvataaarActions";
import Header from "Navigation/Header";
import {isLightThemeStored, setSiteTheme} from "UtilityFunctions";
import * as ProjectRedux from "Project/Redux";

const store = configureStore(initObject(Cloud.homeFolder), {
    activity,
    dashboard,
    analyses,
    applications,
    header,
    status,
    sidebar,
    uploader,
    notifications,
    simpleSearch,
    detailedFileSearch,
    detailedApplicationSearch,
    fileInfo,
    filePreview,
    ...AppRedux.reducers,
    ...AccountingRedux.reducers,
    avatar,
    loading,
    project: ProjectRedux.reducer,
    responsive: createResponsiveStateReducer(
        responsiveBP,
        {infinity: "xxl"}),
}, responsiveStoreEnhancer);

function loading(state = false, action): boolean {
    switch (action.type) {
        case "LOADING_START":
            return true;
        case "LOADING_END":
            return false;
        default:
            return state;
    }
}

export function dispatchUserAction(type: typeof USER_LOGIN | typeof USER_LOGOUT | typeof CONTEXT_SWITCH) {
    store.dispatch({type})
}

export async function onLogin() {
    const action = await findAvatar();
    if (action !== null) store.dispatch(action);
}

const GlobalStyle = createGlobalStyle`
  ${() => UIGlobalStyle}
`;

Cloud.initializeStore(store);

function App({children}) {
    const [isLightTheme, setTheme] = React.useState(isLightThemeStored());
    const setAndStoreTheme = (isLight: boolean) => (setSiteTheme(isLight), setTheme(isLight));
    return (
        <ThemeProvider theme={isLightTheme ? theme : {...theme, colors: invertedColors}}>
            <>
                <GlobalStyle/>
                <BrowserRouter basename="app">
                    <Header toggleTheme={() => isLightTheme ? setAndStoreTheme(false) : setAndStoreTheme(true)}/>
                    {children}
                </BrowserRouter>
            </>
        </ThemeProvider>
    )
}

ReactDOM.render(
    <Provider store={store}>
        <App>
            <Core/>
        </App>
    </Provider>,
    document.getElementById("app")
);
