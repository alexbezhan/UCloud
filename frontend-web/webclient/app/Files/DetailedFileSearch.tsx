import {KeyCode, ReduxObject} from "DefaultObjects";
import * as DFSActions from "Files/Redux/DetailedFileSearchActions";
import {History} from "history";
import * as React from "react";
import {connect} from "react-redux";
import {withRouter} from "react-router";
import {Dispatch} from "redux";
import {snackbarStore} from "Snackbar/SnackbarStore";
import {Button, Checkbox, Flex, Input, InputGroup, Label, OutlineButton, Stamp} from "ui-components";
import Box from "ui-components/Box";
import ClickableDropdown from "ui-components/ClickableDropdown";
import {DatePicker} from "ui-components/DatePicker";
import * as Heading from "ui-components/Heading";
import {searchPage} from "Utilities/SearchUtilities";
import {
    AdvancedSearchRequest,
    DetailedFileSearchOperations,
    DetailedFileSearchReduxState,
    DetailedFileSearchStateProps,
    FileType,
    PossibleTime
} from ".";
import {
    DETAILED_FILES_ADD_EXTENSIONS,
    DETAILED_FILES_ADD_SENSITIVITIES,
    DETAILED_FILES_ADD_TAGS,
    DETAILED_FILES_REMOVE_EXTENSIONS,
    DETAILED_FILES_REMOVE_SENSITIVITIES,
    DETAILED_FILES_REMOVE_TAGS
} from "./Redux/DetailedFileSearchReducer";


interface DetailedFileSearchGivenProps {
    history: History;
    defaultFilename?: string;
    cantHide?: boolean;
    omitFileName?: boolean;
}

type DetailedFileSearchProps = DetailedFileSearchStateProps & DetailedFileSearchGivenProps;

class DetailedFileSearch extends React.Component<DetailedFileSearchProps> {
    private extensionsInput = React.createRef<HTMLInputElement>();

    constructor(props) {
        super(props);
    }

    public componentDidMount() {
        if (!!this.props.defaultFilename) this.props.setFilename(this.props.defaultFilename);
    }


    public componentWillUnmount() {
        if (!this.props.hidden) this.props.toggleHidden();
    }

    public render() {
        const {hidden, cantHide, extensions, allowFiles, allowFolders, includeShares} = this.props;
        if (hidden && !cantHide) {
            return (<OutlineButton fullWidth color="darkGreen" onClick={this.props.toggleHidden}>Advanced
                Search</OutlineButton>);
        }

        return (
            <>
                {!cantHide ? <OutlineButton fullWidth color="darkGreen" onClick={this.props.toggleHidden}>Hide Advanced
                    Search</OutlineButton> : null}
                <Flex flexDirection="column" pl="0.5em" pr="0.5em">
                    <Box mt="0.5em">
                        <form onSubmit={e => (e.preventDefault(), this.onSearch())}>
                            <Heading.h5 pb="0.3em" pt="0.5em">Filename</Heading.h5>
                            <Input
                                pb="6px"
                                pt="8px"
                                mt="-2px"
                                width="100%"
                                placeholder="Filename must include..."
                                value={this.props.fileName}
                                onChange={({target}) => this.props.setFilename(target.value)}
                            />
                            <Heading.h5 pb="0.3em" pt="0.5em">Created at</Heading.h5>
                            <InputGroup>
                                <DatePicker
                                    pb="6px"
                                    pt="8px"
                                    mt="-2px"
                                    placeholderText="After"
                                    selected={this.props.createdAfter}
                                    startDate={this.props.createdAfter}
                                    endDate={this.props.createdBefore}
                                    onChange={d => this.validateAndSetDate(d, "createdAfter")}
                                    showTimeSelect
                                    timeIntervals={15}
                                    isClearable
                                    selectsStart
                                    timeFormat="HH:mm"
                                    dateFormat="dd/MM/yy HH:mm"
                                />
                                <DatePicker
                                    pb="6px"
                                    pt="8px"
                                    mt="-2px"
                                    selectsEnd
                                    placeholderText="Before"
                                    selected={this.props.createdBefore}
                                    startDate={this.props.createdAfter}
                                    endDate={this.props.createdBefore}
                                    onChange={d => this.validateAndSetDate(d, "createdBefore")}
                                    showTimeSelect
                                    timeIntervals={15}
                                    isClearable
                                    timeFormat="HH:mm"
                                    dateFormat="dd/MM/yy HH:mm"
                                />
                            </InputGroup>
                            <Heading.h5 pb="0.3em" pt="0.5em">Modified at</Heading.h5>
                            <InputGroup>
                                <DatePicker
                                    pb="6px"
                                    pt="8px"
                                    mt="-2px"
                                    placeholderText="After"
                                    selected={this.props.modifiedAfter}
                                    selectsStart
                                    startDate={this.props.modifiedAfter}
                                    endDate={this.props.modifiedBefore}
                                    onChange={d => this.validateAndSetDate(d, "modifiedAfter")}
                                    showTimeSelect
                                    timeIntervals={15}
                                    isClearable
                                    timeFormat="HH:mm"
                                    dateFormat="dd/MM/yy HH:mm"
                                />
                                <DatePicker
                                    pb="6px"
                                    pt="8px"
                                    mt="-2px"
                                    placeholderText="Before"
                                    selected={this.props.modifiedBefore}
                                    selectsEnd
                                    startDate={this.props.modifiedAfter}
                                    endDate={this.props.modifiedBefore}
                                    onChange={d => this.validateAndSetDate(d, "modifiedBefore")}
                                    showTimeSelect
                                    timeIntervals={15}
                                    isClearable
                                    timeFormat="HH:mm"
                                    dateFormat="dd/MM/yy HH:mm"
                                />
                            </InputGroup>
                            <Heading.h5 pb="0.3em" pt="0.5em">File Types</Heading.h5>
                            <Flex>
                                <Label fontSize={1} color="black">
                                    <Checkbox
                                        checked={allowFolders}
                                        onChange={(e: React.SyntheticEvent) => e.stopPropagation()}
                                        onClick={() => this.props.toggleFolderAllowed()}
                                    />
                                    Folders
                                </Label>
                                <Label fontSize={1} color="black">
                                    <Checkbox
                                        checked={allowFiles}
                                        onChange={(e: React.SyntheticEvent) => e.stopPropagation()}
                                        onClick={() => this.props.toggleFilesAllowed()}
                                    />
                                    Files
                                </Label>
                            </Flex>
                            <Heading.h5 pb="0.3em" pt="0.5em">Include Shares</Heading.h5>
                            <Flex>
                                <Label fontSize={1} color="black">
                                    <Checkbox
                                        checked={(includeShares)}
                                        onChange={(e: React.SyntheticEvent) => e.stopPropagation()}
                                        onClick={this.props.toggleIncludeShares}
                                    />
                                </Label>
                            </Flex>
                            <Heading.h5 pb="0.3em" pt="0.5em">File extensions</Heading.h5>
                            <SearchStamps
                                stamps={extensions}
                                onStampRemove={l => this.props.removeExtensions([l])}
                                clearAll={() => this.props.removeExtensions([...extensions])}
                            />
                            <Input
                                type="text"
                                pb="6px"
                                pt="8px"
                                mt="-2px"
                                onKeyDown={e => {
                                    if (e.keyCode === KeyCode.ENTER) {
                                        e.preventDefault();
                                        this.onAddExtension();
                                    }
                                }}
                                ref={this.extensionsInput}
                                placeholder={"Add extensions..."}
                            />
                            <ClickableDropdown
                                width="100%"
                                chevron
                                trigger={"Extension presets"}
                                onChange={value => this.onAddPresets(value)}
                                options={extensionPresets}
                            />
                            <Button
                                type="submit"
                                fullWidth
                                disabled={this.props.loading}
                                mt="1em"
                                mb={"1.5em"}
                                color="blue"
                            >Search</Button>
                        </form>
                    </Box>
                </Flex>
            </>
        );
    }

    private onAddExtension() {
        const extensions = this.extensionsInput.current;
        if (!extensions || !extensions.value) return;
        const newExtensions = extensions.value.trim().split(" ").filter(it => it);
        this.props.addExtensions(newExtensions);
        extensions.value = "";
    }

    private onAddPresets(presetExtensions: string) {
        const ext = presetExtensions.trim().split(" ").filter(it => it);
        this.props.addExtensions(ext);
    }

    private validateAndSetDate(m: Date | null, property: PossibleTime) {
        const {setTimes, createdBefore, modifiedBefore, createdAfter, modifiedAfter} = this.props;
        if (m == null) {
            setTimes({[property]: undefined});
            return;
        }
        const before = property.includes("Before");
        if (property.includes("created")) {
            if (before && createdAfter) {
                if (m.getTime() > createdAfter.getTime()) {
                    setTimes({createdBefore: m});
                    return;
                } else {
                    snackbarStore.addFailure("Invalid date range");
                    return;
                }
            } else if (!before && createdBefore) {
                if (m.getTime() < createdBefore.getTime()) {
                    setTimes({createdAfter: m});
                    return;
                } else {
                    snackbarStore.addFailure("Invalid date range");
                    return;
                }
            }
        } else { // includes Modified
            if (before && modifiedAfter) {
                if (m.getTime() > modifiedAfter.getTime()) {
                    setTimes({modifiedBefore: m});
                    return;
                } else {
                    snackbarStore.addFailure("Invalid date range");
                    return;
                }
            } else if (!before && modifiedBefore) {
                if (m.getTime() < modifiedBefore.getTime()) {
                    setTimes({modifiedAfter: m});
                    return;
                } else {
                    snackbarStore.addFailure("Invalid date range");
                    return;
                }
            }
        }
        setTimes({[property]: m});
    }

    private onSearch = () => {
        this.onAddExtension();
        const fileTypes: [FileType?, FileType?] = [];
        if (this.props.allowFiles) fileTypes.push("FILE");
        if (this.props.allowFolders) fileTypes.push("DIRECTORY");
        const createdAt = {
            after: !!this.props.createdAfter ? this.props.createdAfter.valueOf() : undefined,
            before: !!this.props.createdBefore ? this.props.createdBefore.valueOf() : undefined,
        };
        const modifiedAt = {
            after: !!this.props.modifiedAfter ? this.props.modifiedAfter.valueOf() : undefined,
            before: !!this.props.modifiedBefore ? this.props.modifiedBefore.valueOf() : undefined,
        };
        const fileName = this.props.fileName;
        const request: AdvancedSearchRequest = {
            fileName,
            extensions: [...this.props.extensions],
            sensitivity: [...this.props.sensitivities],
            fileTypes,
            createdAt: typeof createdAt.after === "number" || typeof createdAt.before === "number" ?
                createdAt : undefined,
            modifiedAt: typeof modifiedAt.after === "number" || typeof modifiedAt.before === "number" ?
                modifiedAt : undefined,
            itemsPerPage: 25,
            page: 0
        };
        this.props.history.push(searchPage("files", this.props.fileName));
    }
}

interface SearchStampsProps {
    stamps: Set<string>;
    onStampRemove: (stamp: string) => void;
    clearAll: () => void;
}

const SearchStamps = ({stamps, onStampRemove, clearAll}: SearchStampsProps) => (
    <Box pb="5px">
        {[...stamps].map(l => (
            <Stamp onClick={() => onStampRemove(l)} ml="2px" mt="2px" color="blue" key={l} text={l} />))}
        {stamps.size > 1 ? (<Stamp ml="2px" mt="2px" color="red" onClick={() => clearAll()} text="Clear all" />) : null}
    </Box>
);

const extensionPresets = [
    {text: "Text", value: ".txt .docx .rtf .csv .pdf"},
    {text: "Image", value: ".png .jpeg .jpg .ppm .gif"},
    {text: "Sound", value: ".mp3 .ogg .wav .flac .aac"},
    {text: "Compressed files", value: ".zip .tar.gz"}
];

const mapStateToProps = ({detailedFileSearch}: ReduxObject): DetailedFileSearchReduxState & {sizeCount: number} => ({
    ...detailedFileSearch,
    sizeCount: detailedFileSearch.extensions.size + detailedFileSearch.tags.size + detailedFileSearch.sensitivities.size
});


const mapDispatchToProps = (dispatch: Dispatch): DetailedFileSearchOperations => ({
    toggleHidden: () => dispatch(DFSActions.toggleFilesSearchHidden()),
    addExtensions: ext => dispatch(DFSActions.extensionAction(DETAILED_FILES_ADD_EXTENSIONS, ext)),
    removeExtensions: ext => dispatch(DFSActions.extensionAction(DETAILED_FILES_REMOVE_EXTENSIONS, ext)),
    toggleFolderAllowed: () => dispatch(DFSActions.toggleFoldersAllowed()),
    toggleFilesAllowed: () => dispatch(DFSActions.toggleFilesAllowed()),
    toggleIncludeShares: () => dispatch(DFSActions.toggleIncludeShares()),
    addSensitivity: sens => dispatch(DFSActions.sensitivityAction(DETAILED_FILES_ADD_SENSITIVITIES, [sens])),
    removeSensitivity: sens => dispatch(DFSActions.sensitivityAction(DETAILED_FILES_REMOVE_SENSITIVITIES, sens)),
    addTags: tags => dispatch(DFSActions.tagAction(DETAILED_FILES_ADD_TAGS, tags)),
    removeTags: tags => dispatch(DFSActions.tagAction(DETAILED_FILES_REMOVE_TAGS, tags)),
    setFilename: filename => dispatch(DFSActions.setFilename(filename)),
    setLoading: loading => dispatch(DFSActions.setFilesSearchLoading(loading)),
    setTimes: times => dispatch(DFSActions.setTime(times)),
});

export default connect(mapStateToProps, mapDispatchToProps)(withRouter(DetailedFileSearch));
