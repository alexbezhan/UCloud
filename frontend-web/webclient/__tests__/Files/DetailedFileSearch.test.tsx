import * as React from "react";
import DetailedFileSearch from "../../app/Files/DetailedFileSearch";
import { Provider } from "react-redux";
import { configureStore } from "../../app/Utilities/ReduxUtilities";
import { initObject, initFilesDetailedSearch } from "../../app/DefaultObjects";
import detailedFileSearch from "../../app/Files/Redux/DetailedFileSearchReducer";
import { configure } from "enzyme";
import * as Adapter from "enzyme-adapter-react-16";
import "jest-styled-components";
import { ThemeProvider } from "styled-components";
import theme from "../../app/ui-components/theme";
import { create } from "react-test-renderer";
import { MemoryRouter } from "react-router";

configure({ adapter: new Adapter() });

const store = configureStore({ detailedFileSearch: initFilesDetailedSearch() }, { detailedFileSearch });

describe("Detailed File Search", () => {
    it("Mount file search", () => {
        expect(create(<Provider store={store}>
            <ThemeProvider theme={theme}>
                <MemoryRouter>
                    <DetailedFileSearch

                    />
                </MemoryRouter>
            </ThemeProvider>
        </Provider>).toJSON()).toMatchSnapshot()
    });
})