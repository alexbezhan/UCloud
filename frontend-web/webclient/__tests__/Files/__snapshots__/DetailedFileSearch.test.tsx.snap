// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Detailed File Search Mount file search 1`] = `
.c0 {
  -webkit-font-smoothing: antialiased;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  text-align: center;
  -webkit-text-decoration: none;
  text-decoration: none;
  font-family: inherit;
  font-weight: 700;
  line-height: 1.5;
  cursor: pointer;
  border-radius: 5px;
  background-color: #00823c;
  color: #fff;
  border-width: 0;
  border-style: solid;
  -webkit-transition: cubic-bezier(0.5,0,0.25,1) 60ms;
  transition: cubic-bezier(0.5,0,0.25,1) 60ms;
  width: 100%;
  font-size: 14px;
  padding: 9.5px 18px;
  color: #00823c;
  border: 2px solid #00823c;
  border-radius: 5px;
  background-color: transparent;
  font-size: 14px;
  padding: 7.5px 18px;
}

.c0:disabled {
  opacity: 0.25;
}

.c0:focus {
  outline: none;
}

.c0:hover {
  -webkit-transform: scale(1.03);
  -ms-transform: scale(1.03);
  transform: scale(1.03);
}

.c0:hover {
  border: 2px solid;
  background-color: transparent;
  -webkit-transition: ease 0.1s;
  transition: ease 0.1s;
}

<button
  className="c0"
  color="darkGreen"
  onClick={[Function]}
>
  Advanced Search
</button>
`;
